dlib/image_transforms/interpolation.h

```diff
@@ -965,6 +965,46 @@ namespace dlib
         swap(img, temp);
     }
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename image_type>
+    point_transform_affine letterbox_image (
+        const image_type& img_in,
+        image_type& img_out,
+        long size
+    )
+    {
+        DLIB_CASSERT(size > 0, "size must be bigger than zero, but was " << size);
+        // the scaling factor of the image
+        const auto scale = size / std::max<double>(img_in.nr(), img_in.nc());
+
+        // early return if the image has already the requested size and no padding is needed
+        if (scale == 1 and img_in.nr() == img_in.nc())
+        {
+            img_out = img_in;
+            return point_transform_affine();
+        }
+
+        // black background
+        img_out.set_size(size, size);
+        assign_all_pixels(img_out, 0);
+
+        // resize the image so that it fits into a size x size image
+        image_type temp = img_in;
+        resize_image(scale, temp);
+
+        // get the row and column offsets (the padding size)
+        const point offset((size - temp.nc()) / 2, (size - temp.nr()) / 2);
+        for (long r = 0; r < temp.nr(); ++r)
+        {
+            for (long c = 0; c < temp.nc(); ++c)
+            {
+                img_out(offset.y() + r, offset.x() + c) = temp(r, c);
+            }
+        }
+        return point_transform_affine(identity_matrix<double>(2) * scale, offset);
```

Comment:
This whole function could be done with one call to `transform_image()` if you supplied the appropriate `point_transform_affine` as the `map_point`.  Are you doing it this way because it's faster?  Like because it will hit the simd optimized `resize_image()`?  What's here is making multiple passes at the image, which is not good for performance, but maybe `resize_image()` is so fast it wins out?

Whatever is faster is fine.  But you should at least replace the for loops here with an `assign_image()` statement that uses `sub_image()`.  Like `assign_image(sub_image(img_out, some_rect), temp)`.